### PR TITLE
Feature/wa 271 setup recovery screen

### PR DESCRIPTION
**Pull request by sche**
Introduce root view controller for Flow coordinators for more control;
Refactor OnboardingFlowCoordinator and MasterPasswordFlowCoordinator;

### CODE REVIEW

File: safe/safe/FlowCoordinator.swift [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import UIKit
+
+class FlowCoordinator {
+
+    var rootVC: UINavigationController!
+
+    func startViewController(parent: UINavigationController? = nil) -> UIViewController {
+        // rootVC should be set before caling flowStartController
+        rootVC = parent != nil ? parent : TransparentNavigationController()
```
Comment (DmitryBespalov) [inline safe/safe/FlowCoordinator.swift]: I would reverse boolean condition for readability: `parent == nil ? YES_CASE : NO_CASE`

File: safe/safe/FlowCoordinator.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+//
+
+import UIKit
+
+class FlowCoordinator {
+
+    var rootVC: UINavigationController!
+
+    func startViewController(parent: UINavigationController? = nil) -> UIViewController {
+        // rootVC should be set before caling flowStartController
```
Comment (DmitryBespalov) [inline safe/safe/FlowCoordinator.swift]: imo this comment is not needed

File: safe/safe/FlowCoordinatorTests.swift [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        XCTAssertTrue(flowCoordinator.rootViewControllerIsSet)
+        XCTAssertTrue(flowCoordinator.rootVC is TransparentNavigationController)
+    }
+
+    func test_startViewController_setsRootIfParentPassed() {
+        let parent = UINavigationController()
+        _ = flowCoordinator.startViewController(parent: parent)
+        XCTAssertEqual(parent, flowCoordinator.rootVC)
+    }
+
```
Comment (DmitryBespalov) [inline safe/safe/FlowCoordinatorTests.swift:24]: missing test for rootVC.pushViewController - line 16 in FlowCoordinator

File: safe/safe/OnboardingFlow/MasterPasswordFlow/MasterPasswordFlowCoordinator.swift [hunk 4]
```diff
@@ -21,7 +21,8 @@ extension MasterPasswordFlowCoordinator: StartViewControllerDelegate {
 
     func didStart() {
         let vc = SetPasswordViewController.create(delegate: self)
-        masterPasswordNavigationController.show(vc, sender: nil)
+        print(rootVC)
```
Comment (DmitryBespalov) [inline safe/safe/OnboardingFlow/MasterPasswordFlow/MasterPasswordFlowCoordinator.swift]: debug print should be avoided in production code

**Review by DmitryBespalov — CHANGES_REQUESTED** [review]
some refactoring needed

File: safe/safe/FlowCoordinator.swift (hunk 1 above)
Comment (sche) [reply safe/safe/FlowCoordinator.swift]: ok

File: safe/safe/FlowCoordinator.swift (hunk 2 above)
Comment (sche) [reply safe/safe/FlowCoordinator.swift]: ok

**Review by DmitryBespalov — APPROVED** [review]

File: safe/safe/OnboardingFlow/OnboardingFlowCoordinator.swift [hunk 5]
```diff
@@ -4,19 +4,27 @@
...
     private let account: AccountProtocol
-    let masterPasswordFlowCoordinator = MasterPasswordFlowCoordinator()
+    let masterPasswordFlowCoordinator: MasterPasswordFlowCoordinator
     let setupSafeFlowCoordinator = SetupSafeFlowCoordinator()
 
     init(account: AccountProtocol) {
         self.account = account
+        masterPasswordFlowCoordinator = MasterPasswordFlowCoordinator()
+        super.init()
+        masterPasswordFlowCoordinator.completion = masterPasswordCompletion
```
Comment (DmitryBespalov) [inline safe/safe/OnboardingFlow/OnboardingFlowCoordinator.swift]: 👍

**Review by DmitryBespalov — APPROVED** [review]

**Pull request merged**